### PR TITLE
Neural Network Extension, HIP GPU backend - add support for layers that use MIOpen

### PR DESCRIPTION
**Pull request by asalmanp**
- This PR adds support for NN layers that call MIOpen. These include convolution, deconvolution, fully_connected, 
local_response_normalization, normalization, pooling, reshape, scale, softmax, tensor_add, tensor_max, tensor_min, tensor_multiply, 
 and tensor_subtract layers.

- This PR has partial support for tensor_matrix_multiply. We would need to use rocBLAS for doing GEMM which will be added in the next PR for this layer.

### CODE REVIEW

File: amd_openvx_extensions/amd_nn/src/batch_normalization_layer.cpp [hunk 1]
```diff
@@ -25,15 +25,28 @@ THE SOFTWARE.
...
     cl_mem input_mem;
+#elif ENABLE_HIP
+    vx_uint8 *input_mem;
+#endif
     miopenTensorDescriptor_t output_desc;
     miopenDataType_t data_type;          // data_type for the kernel
+#if ENABLE_OPENCL
     cl_mem output_mem;
     cl_mem workspace;
+#elif ENABLE_HIP
```
Comment (rrawther) [inline amd_openvx_extensions/amd_nn/src/batch_normalization_layer.cpp:38]: if we keep this (void*), we can use the same variables for both HIP and OpenCL. Only cast it where it is used

File: amd_openvx_extensions/amd_nn/src/batch_normalization_layer.cpp [hunk 2]
```diff
@@ -25,15 +25,28 @@ THE SOFTWARE.
...
+    vx_uint8 *output_mem;
+    vx_uint8 *workspace;
+#endif
     size_t workspace_size;
     float alpha, beta, eps;
     miopenTensorDescriptor_t bnScaleBiasMeanVarDesc;
+#if ENABLE_OPENCL
```
Comment (rrawther) [inline amd_openvx_extensions/amd_nn/src/batch_normalization_layer.cpp:45]: same comment as before

File: amd_openvx_extensions/amd_nn/src/batch_normalization_layer.cpp [hunk 3]
```diff
@@ -172,6 +192,40 @@ static vx_status VX_CALLBACK initializeBatchNormalizationLayer(vx_node node, con
...
+        ERROR_CHECK_STATUS(vxQueryTensor((vx_tensor)parameters[4], VX_TENSOR_BUFFER_HIP, &data->bnBias, sizeof(data->bnBias)));
+    } else {
+        vx_context vxContext = vxGetContext((vx_reference)node);
+        int hip_device = -1;
+        ERROR_CHECK_STATUS(vxQueryContext(vxContext, VX_CONTEXT_ATTRIBUTE_AMD_HIP_DEVICE, &hip_device, sizeof(hip_device)));
+        if (hip_device < 0) {
+            return VX_FAILURE;
+        }
+        hipError_t errcode_ret = hipSuccess;
+        errcode_ret = hipSetDevice(hip_device);
```
Comment (rrawther) [inline amd_openvx_extensions/amd_nn/src/batch_normalization_layer.cpp:211]: Do we need to do this again here since hip_device is already set for the context

File: amd_openvx_extensions/amd_nn/src/convolution_layer.cpp [hunk 4]
```diff
@@ -370,6 +408,29 @@ static vx_status VX_CALLBACK initializeConvolutionLayer(vx_node node, const vx_r
...
                 err = clEnqueueFillBuffer(data->handle->cmdq, data->workspace, &pattern, sizeof(cl_half), 0, data->workspace_size, 0, NULL, NULL);
             if(err) return VX_FAILURE;
+#elif ENABLE_HIP
+            int hip_device = -1;
+            ERROR_CHECK_STATUS(vxQueryContext(vxContext, VX_CONTEXT_ATTRIBUTE_AMD_HIP_DEVICE, &hip_device, sizeof(hip_device)));
+            if (hip_device < 0) {
+                return VX_FAILURE;
+            }
+            hipError_t errcode_ret = hipSuccess;
+            errcode_ret = hipSetDevice(hip_device);
```
Comment (rrawther) [inline amd_openvx_extensions/amd_nn/src/convolution_layer.cpp:418]: do we need to set the device again

File: amd_openvx_extensions/amd_nn/src/kernels.h [hunk 5]
```diff
@@ -48,7 +48,12 @@ THE SOFTWARE.
 #if __APPLE__
 #include <opencl.h>
 #else
+#if ENABLE_OPENCL
 #include <CL/cl.h>
+#elif ENABLE_HIP
+#include  "../../../amd_openvx/openvx/ago/ago_internal.h"
```
Comment (rrawther) [inline amd_openvx_extensions/amd_nn/src/kernels.h:54]: It is not a good idea to include "ago_internal.h" here. Why is this required?

File: amd_openvx_extensions/amd_nn/src/scale_layer.cpp [hunk 6]
```diff
@@ -159,8 +185,46 @@ static vx_status VX_CALLBACK initializeScaleLayer(vx_node node, const vx_referen
...
         }
         if (err) return VX_FAILURE;
+#elif ENABLE_HIP
+        int hip_device = -1;
+        ERROR_CHECK_STATUS(vxQueryContext(vxContext, VX_CONTEXT_ATTRIBUTE_AMD_HIP_DEVICE, &hip_device, sizeof(hip_device)));
+        if (hip_device < 0) {
+            return VX_FAILURE;
+        }
+        hipError_t errcode_ret = hipSuccess;
+        errcode_ret = hipSetDevice(hip_device);
```
Comment (rrawther) [inline amd_openvx_extensions/amd_nn/src/scale_layer.cpp:195]: why we need to set device again here?

**Review by rrawther — CHANGES_REQUESTED** [review]
@paveltc to run all the tests

File: amd_openvx_extensions/amd_nn/src/batch_normalization_layer.cpp (hunk 3 above)
Comment (asalmanp) [reply amd_openvx_extensions/amd_nn/src/batch_normalization_layer.cpp]: No, we don't need it. I removed it.

File: amd_openvx_extensions/amd_nn/src/convolution_layer.cpp (hunk 4 above)
Comment (asalmanp) [reply amd_openvx_extensions/amd_nn/src/convolution_layer.cpp]: removed it.

File: amd_openvx_extensions/amd_nn/src/scale_layer.cpp (hunk 6 above)
Comment (asalmanp) [reply amd_openvx_extensions/amd_nn/src/scale_layer.cpp]: removed it.

File: amd_openvx_extensions/amd_nn/src/kernels.h (hunk 5 above)
Comment (asalmanp) [reply amd_openvx_extensions/amd_nn/src/kernels.h:54]: we need to have access to AgoData (i.e, it is used to cast the [parameters](https://github.com/GPUOpen-ProfessionalCompute-Libraries/MIVisionX/blob/master/amd_openvx_extensions/amd_nn/src/gather_layer.cpp#L218) for custom HIP kernel launch) which is defined in ago_internal.h.

File: amd_openvx_extensions/amd_nn/src/batch_normalization_layer.cpp (hunk 1 above)
Comment (asalmanp) [reply amd_openvx_extensions/amd_nn/src/batch_normalization_layer.cpp:38]: Changing the type of the variable to void* would require changing the call for OCL APIs as well which is not related to this PR. I would suggest keeping the current changes. We can make that change in a separate PR.

**Comment by paveltc** [comment]
@kiritigowda @asalmanp @rrawther Which version of MIOpen is needed to run everything in this PR?

**Comment by kiritigowda** [comment]
> @kiritigowda @asalmanp @rrawther Which version of MIOpen is needed to run everything in this PR?

Can you use the V2.11.0, this is what we are providing with current setup script

**Comment by asalmanp** [comment]
> > @kiritigowda @asalmanp @rrawther Which version of MIOpen is needed to run everything in this PR?
> 
> Can you use the V2.11.0, this is what we are providing with current setup script

@paveltc you need to run the MIVisionX-setup.py --backend=hip to install the miopen with the hip backend before building the mivisionx with the hip backend for testing this PR.

**Comment by paveltc** [comment]
@asalmanp @kiritigowda I sent an email on Friday where I listed the errors that I got while running the setup script. Any updates on that?

File: amd_openvx_extensions/amd_nn/src/batch_normalization_layer.cpp (hunk 1 above)
Comment (rrawther) [reply amd_openvx_extensions/amd_nn/src/batch_normalization_layer.cpp:38]: If you want to keep this format, I would suggest keep all members in the structure within a single #if, #else.

File: amd_openvx_extensions/amd_nn/src/kernels.h (hunk 5 above)
Comment (rrawther) [reply amd_openvx_extensions/amd_nn/src/kernels.h:54]: We need to find another way. AgoData can only be used with amdovx project. All other projects need to interface through vxapi.h or vx_amd_ext.h.

**Review by rrawther — CHANGES_REQUESTED** [review]
please review the comments in the PR

File: amd_openvx_extensions/amd_nn/src/batch_normalization_layer.cpp (hunk 1 above)
Comment (asalmanp) [reply amd_openvx_extensions/amd_nn/src/batch_normalization_layer.cpp]: switched to void* for LocalData.

File: amd_openvx_extensions/amd_nn/src/kernels.h (hunk 5 above)
Comment (asalmanp) [reply amd_openvx_extensions/amd_nn/src/kernels.h]: removed including "ago_internal.h"

**Review by rrawther — APPROVED** [review]

**Comment by kiritigowda** [comment]
@paveltc can this PR be merged?

**Comment by paveltc** [comment]
@kiritigowda Working on running the tests now.

**Comment by paveltc** [comment]
**Unit test results**
Batch Normalization: PASS
Local Response Normalization: PASS
Softmax: PASS
Tensor Add: PASS
Tensor Multiply: PASS
Tensor Subtract: PASS
Pooling: PASS
Scale: PASS
Convolution: Small mismatches between HIP and OCL layer results
Fully Connected: Small mismatches between HIP and OCL layer results
Deconvolution: Segmentation fault with OCL on PR619 branch, no segmentation fault on TOT. Mismatches between TOT OCL and HIP PR619.

PASS = OpenCL and HIP results match

**Comment by paveltc** [comment]
@asalmanp 
Tests update:
Reshape: PASS
Convolution: still small differences between HIP and OCL
Deconvolution: small differences between HIP and OCL. No more segmentation fault
Fully Connected: still small differences between HIP and OCL

**Comment by paveltc** [comment]
Per @asalmanp request, adding a sampling of the differences for the layers mentioned.
**Deconvolution**
First bytes
OCL
0000000       -0.819574      -1.2741065      -1.5055298      -1.7213209
HIP
0000000       -0.819574      -1.2741066      -1.5055298      -1.7213209

Last bytes
OCL
0164260    -0.010193704     0.002807483    -0.013146199  -0.00017771152
HIP
0164260    -0.010193703    0.0028074845      -0.0131462  -0.00017771141
-------------------------------------------------------------------------------------------
The rest of the examples only show first bytes of outputs

**Convolution** 
OCL
0000000     -0.14978679     -0.14978679     -0.14978679     -0.14978679
HIP
0000000     -0.14978677     -0.14978677     -0.14978677     -0.14978677

**Fully Connected**
OCL
0000000      -0.2192885      0.44633734      -1.2363861      -0.6913732
HIP
0000000     -0.21928841      0.44633722      -1.2363861      -0.6913732

**Comment by asalmanp** [comment]
> Per @asalmanp request, adding a sampling of the differences for the layers mentioned.
> **Deconvolution**
> First bytes
> OCL
> 0000000 -0.819574 -1.2741065 -1.5055298 -1.7213209
> HIP
> 0000000 -0.819574 -1.2741066 -1.5055298 -1.7213209
> 
> Last bytes
> OCL
> 0164260 -0.010193704 0.002807483 -0.013146199 -0.00017771152
> HIP
> 0164260 -0.010193703 0.0028074845 -0.0131462 -0.00017771141
> 
> The rest of the examples only show first bytes of outputs
> 
> **Convolution**
> OCL
> 0000000 -0.14978679 -0.14978679 -0.14978679 -0.14978679
> HIP
> 0000000 -0.14978677 -0.14978677 -0.14978677 -0.14978677
> 
> **Fully Connected**
> OCL
> 0000000 -0.2192885 0.44633734 -1.2363861 -0.6913732
> HIP
> 0000000 -0.21928841 0.44633722 -1.2363861 -0.6913732

@paveltc Thanks for providing the result. As discussed, the differences are in the range of 1e-7 or less. This should be fine.
